reconcile/utils/merge_request_manager/merge_request_manager.py

```diff
@@ -0,0 +1,104 @@
+
+
+T = TypeVar("T", bound=OpenMergeRequest)
+
+
+class MergeRequestManagerBase(Generic[T]):
+    """ """
+
+    def __init__(self, vcs: VCS, parser: Parser, mr_label: str, omr_klass: type[T]):
+        self._vcs = vcs
+        self._parser = parser
+        self._mr_label = mr_label
+        self._omr_klass = omr_klass
+        self._open_mrs: list[T] = []
+        self._open_mrs_with_problems: list[T] = []
+        self._housekeeping_ran = False
+
+    @abstractmethod
+    def create_merge_request(self, data: Any) -> None:
+        pass
+
+    def _merge_request_already_exists(
+        self,
+        expected_data: dict[str, Any],
+    ) -> T | None:
+        for mr in self._open_mrs:
+            mr_info_dict = mr.mr_info.dict()
+            if all(
+                mr_info_dict.get(k) == expected_data.get(k)
+                for k in expected_data.keys()
+            ):
+                return mr
+
+        return None
+
+    def _fetch_managed_open_merge_requests(self) -> list[ProjectMergeRequest]:
+        all_open_mrs = self._vcs.get_open_app_interface_merge_requests()
+        return [mr for mr in all_open_mrs if self._mr_label in mr.labels]
+
+    def housekeeping(self) -> None:
+        """
+        Close bad MRs:
+        - bad description format
+        - wrong version
+        - merge conflict
+
+        --> if we update the template output, we automatically close
+        old open MRs and replace them with new ones.
+        """
+        for mr in self._fetch_managed_open_merge_requests():
+            attrs = mr.attributes
+            desc = attrs.get("description")
+            has_conflicts = attrs.get("has_conflicts", False)
+            if has_conflicts:
+                logging.info(
+                    "Merge-conflict detected. Closing %s",
+                    mr.attributes.get("web_url", "NO_WEBURL"),
+                )
+                self._vcs.close_app_interface_mr(
+                    mr, "Closing this MR because of a merge-conflict."
+                )
+                continue
+            try:
+                mr_info = self._parser.parse(description=desc)
+            except ParserVersionError:
+                logging.info(
+                    "Old MR version detected! Closing %s",
+                    mr.attributes.get("web_url", "NO_WEBURL"),
+                )
+                self._vcs.close_app_interface_mr(
+                    mr, "Closing this MR because it has an outdated integration version"
+                )
+                continue
+            except ParserError:
+                logging.info(
+                    "Bad MR description format. Closing %s",
+                    mr.attributes.get("web_url", "NO_WEBURL"),
+                )
+                self._vcs.close_app_interface_mr(
+                    mr, "Closing this MR because of bad description format."
+                )
+                continue
+            self._open_mrs.append(self._omr_klass(raw=mr, mr_info=mr_info))
```

Comment:
you don't need to pass `omr_klass` if use generic all the way down. Just use one template type var for the different part.

```python
T = TypeVar("T", bound=BaseModel)

@dataclass
class OpenMergeRequest(Generic[T]):
    raw: ProjectMergeRequest
    mr_info: T

class MergeRequestManagerBase(Generic[T]):
    """ """

    def __init__(self, vcs: VCS, parser: Parser[T], mr_label: str):
        self._vcs = vcs
        self._parser = parser
        self._mr_label = mr_label
        self._open_mrs: list[OpenMergeRequest[T]] = []
        self._open_mrs_with_problems: list[OpenMergeRequest[T]] = []
        self._housekeeping_ran = False
```

Then this line can be

```python
self._open_mrs.append(OpenMergeRequest[T](raw=mr, mr_info=mr_info))
```

no need to have another layer of

```python
class MRInfo(OpenMergeRequest):
    mr_info: AVSInfo
```

directly use 
```python
class MergeRequestManager(MergeRequestManagerBase[AVSInfo]):
```

Reply:
Sure can do as well. I find this btw way more complicated than a single-line type assertion :-D

Reply:
it's all for mypy check, runtime won't have generic type info at all...